File: src/ragmax/api/v1/evaluation.py
"""Evaluation API endpoints."""

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from ragmax.api.auth_dependencies import ROUTE_EVALUATION, require_route_permission
from ragmax.evaluation.models import TestDataset
from ragmax.evaluation.repository import EvaluationRepository
from ragmax.infrastructure.db.session import get_db_session

router = APIRouter(
    prefix="/evaluation",
    tags=["evaluation"],
    dependencies=[Depends(require_route_permission(ROUTE_EVALUATION))],
)


# ==================== Dataset Endpoints ====================


@router.get("/datasets")
async def list_datasets(
    limit: int = 100,
    db: Annotated[AsyncSession, Depends(get_db_session)] = None,
):
    """List all test datasets."""
    repo = EvaluationRepository(db)
    datasets = await repo.list_datasets(limit=limit)

    return {
        "datasets": [
            {
                "id": ds.id,
                "name": ds.name,
                "description": ds.description,
                "version": ds.version,
                "test_case_count": len(ds.test_cases),
                "created_at": ds.created_at.isoformat(),
            }
            for ds in datasets
        ]
    }


@router.get("/datasets/{dataset_id}")
async def get_dataset(
    dataset_id: str,
    db: Annotated[AsyncSession, Depends(get_db_session)] = None,
):
    """Get a specific dataset with all test cases."""
    repo = EvaluationRepository(db)
    dataset = await repo.get_dataset(dataset_id)

    if not dataset:
        raise HTTPException(status_code=404, detail="Dataset not found")

    return {
        "id": dataset.id,
        "name": dataset.name,
        "description": dataset.description,
        "version": dataset.version,
        "created_at": dataset.created_at.isoformat(),
        "test_cases": [
            {
                "id": tc.id,
                "question": tc.question,
                "expected_answer": tc.expected_answer,
                "ground_truth_docs": tc.ground_truth_docs,
                "metadata": tc.metadata,
            }
            for tc in dataset.test_cases
        ],
    }


@router.post("/datasets")
async def create_dataset(
    dataset: dict,
    db: Annotated[AsyncSession, Depends(get_db_session)] = None,
):
    """Create a new test dataset."""
    from datetime import datetime
    from uuid import uuid4

    from ragmax.evaluation.models import EvalTestCase

    # Parse dataset
    test_cases = [
        EvalTestCase(
            id=tc.get("id", f"tc_{uuid4().hex[:8]}"),
            question=tc["question"],
            expected_answer=tc.get("expected_answer"),
            ground_truth_docs=tc.get("ground_truth_docs", []),
            metadata=tc.get("metadata", {}),
            created_at=datetime.now(),
        )
        for tc in dataset.get("test_cases", [])
    ]

    new_dataset = TestDataset(
        id=str(uuid4()),
        name=dataset["name"],
        description=dataset.get("description", ""),
        test_cases=test_cases,
        version=dataset.get("version", "1.0.0"),
        created_at=datetime.now(),
    )

    repo = EvaluationRepository(db)
    await repo.create_dataset(new_dataset)

    return {"id": new_dataset.id, "message": "Dataset created successfully"}


@router.delete("/datasets/{dataset_id}")
async def delete_dataset(
    dataset_id: str,
    db: Annotated[AsyncSession, Depends(get_db_session)] = None,
):
    """Delete a dataset."""
    repo = EvaluationRepository(db)
    await repo.delete_dataset(dataset_id)
    return {"message": "Dataset deleted successfully"}


# ==================== Experiment Endpoints ====================


@router.get("/experiments")
async def list_experiments(
    dataset_id: str | None = None,
    limit: int = 50,
    db: Annotated[AsyncSession, Depends(get_db_session)] = None,
):
    """List evaluation experiments."""
    repo = EvaluationRepository(db)
    await repo.list_experiments(dataset_id=dataset_id, limit=limit)

    # For now return empty since we need to implement full reconstruction
    # TODO: Implement full experiment loading in repository
    return {"experiments": []}


@router.get("/experiments/{experiment_id}")
async def get_experiment(
    experiment_id: str,
    db: Annotated[AsyncSession, Depends(get_db_session)] = None,
):
    """Get experiment details with results."""
    repo = EvaluationRepository(db)
    experiment = await repo.get_experiment(experiment_id)

    if not experiment:
        raise HTTPException(status_code=404, detail="Experiment not found")

    # TODO: Implement full serialization
    return {"experiment": None}


@router.post("/experiments/run")
async def run_experiment(
    request: dict,
    db: Annotated[AsyncSession, Depends(get_db_session)] = None,
):
    """Run a new evaluation experiment."""
    # TODO: Implement experiment execution
    # This will require initializing RAGEvaluator with proper dependencies

    return {
        "message": "Experiment execution not yet implemented",
        "status": "pending",
    }


# ==================== Version Management ====================


@router.get("/datasets/{dataset_name}/versions")
async def list_dataset_versions(
    dataset_name: str,
    db: Annotated[AsyncSession, Depends(get_db_session)] = None,
):
    """List all versions of a dataset."""
    repo = EvaluationRepository(db)
    versions = await repo.list_dataset_versions(dataset_name)

    return {
        "versions": [
            {
                "id": ds.id,
                "version": ds.version,
                "created_at": ds.created_at.isoformat(),
            }
            for ds in versions
        ]
    }


@router.get("/datasets/{dataset_name}/compare")
async def compare_dataset_versions(
    dataset_name: str,
    version_a: str,
    version_b: str,
    db: Annotated[AsyncSession, Depends(get_db_session)] = None,
):
    """Compare two versions of a dataset."""
    repo = EvaluationRepository(db)
    comparison = await repo.compare_dataset_versions(dataset_name, version_a, version_b)

    return comparison

File: web/src/api/evaluation.ts
/**
 * Evaluation API client
 */

import { apiBaseUrl, authenticatedFetch, parseJsonResponse } from './client';

export interface TestCase {
  id: string;
  question: string;
  expected_answer?: string;
  ground_truth_docs: string[];
  metadata: Record<string, unknown>;
}

export interface Dataset {
  id: string;
  name: string;
  description: string;
  version: string;
  test_case_count?: number;
  test_cases?: TestCase[];
  created_at: string;
}

export interface MetricsSummary {
  context_precision: number;
  context_recall: number;
  faithfulness: number;
  answer_relevancy: number;
  overall_score: number;
  e2e_latency_p95: number;
  pass_rate: number;
  retrieval_latency_p95: number;
  avg_token_cost: number;
}

export interface Experiment {
  id: string;
  name: string;
  dataset_id: string;
  status: 'pending' | 'running' | 'completed' | 'failed';
  started_at: string;
  completed_at?: string;
  duration_seconds?: number;
  metrics_summary?: MetricsSummary;
  config: Record<string, unknown>;
}

export interface DatasetVersion {
  id: string;
  version: string;
  created_at: string;
}

export interface VersionComparison {
  added: number;
  removed: number;
  modified: number;
  unchanged: number;
  total_a: number;
  total_b: number;
}

export const evaluationApi = {
  // Dataset operations
  async listDatasets(limit = 100): Promise<Dataset[]> {
    const response = await requestJson<{ datasets: Dataset[] }>(
      `/api/v1/evaluation/datasets?limit=${limit}`
    );
    return response.datasets;
  },

  async getDataset(datasetId: string): Promise<Dataset> {
    return requestJson(`/api/v1/evaluation/datasets/${encodeURIComponent(datasetId)}`);
  },

  async createDataset(dataset: {
    name: string;
    description?: string;
    version?: string;
    test_cases: Partial<TestCase>[];
  }): Promise<{ id: string; message: string }> {
    return requestJson('/api/v1/evaluation/datasets', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(dataset),
    });
  },

  async deleteDataset(datasetId: string): Promise<{ message: string }> {
    return requestJson(`/api/v1/evaluation/datasets/${encodeURIComponent(datasetId)}`, {
      method: 'DELETE',
    });
  },

  // Experiment operations
  async listExperiments(datasetId?: string, limit = 50): Promise<Experiment[]> {
    const params = new URLSearchParams();
    if (datasetId) params.append('dataset_id', datasetId);
    params.append('limit', limit.toString());

    const response = await requestJson<{ experiments: Experiment[] }>(
      `/api/v1/evaluation/experiments?${params}`
    );
    return response.experiments;
  },

  async getExperiment(experimentId: string): Promise<Experiment | null> {
    const response = await requestJson<{ experiment: Experiment | null }>(
      `/api/v1/evaluation/experiments/${encodeURIComponent(experimentId)}`
    );
    return response.experiment;
  },

  async runExperiment(request: {
    dataset_id: string;
    name: string;
    config: Record<string, unknown>;
  }): Promise<{ message: string; status: string }> {
    return requestJson('/api/v1/evaluation/experiments/run', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });
  },

  // Version management
  async listDatasetVersions(datasetName: string): Promise<DatasetVersion[]> {
    const response = await requestJson<{ versions: DatasetVersion[] }>(
      `/api/v1/evaluation/datasets/${encodeURIComponent(datasetName)}/versions`
    );
    return response.versions;
  },

  async compareDatasetVersions(
    datasetName: string,
    versionA: string,
    versionB: string
  ): Promise<VersionComparison> {
    const params = new URLSearchParams({
      version_a: versionA,
      version_b: versionB,
    });
    return requestJson(
      `/api/v1/evaluation/datasets/${encodeURIComponent(datasetName)}/compare?${params}`
    );
  },
};

async function requestJson<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await authenticatedFetch(`${apiBaseUrl}${path}`, init);
  return parseJsonResponse<T>(response);
}
